### DISCUSSION TITLE
How do I start my app, and have my libraries build before the app starts?

### DISCUSSION THREAD

**Discussion by NullVoxPopuli** [vercel/turborepo#372] [Help] [answered]
My monorepo is fairly simple(tm):

```
 - app
    - yarn start (boots dev server)
    
 - packages/
     - package-a/
        - yarn build
     - package-b/
        - yarn build
     - others
```

the app depends on both packages.

my turbo config is 
```json
{
  "turbo": {
    "pipeline": {
      "start": { "dependsOn": "^build" }
    }
  }
}
```        

however, when I run:
```js
yarn turbo run start --scope="app"
```
the build commands from the packages are not run.

I also tried 
```
yarn turbo run start --cwd="./app"
```
but it tried to run start on _everything_ (non-relevant packages, apps, etc).. which I don't want. I just want to develop one app atm.

how can I do this?

Thanks!!

**Comment by mmJordanO** [comment]
I have a similar situation and would like to see an example of this.  If I run my app's dev server, is there a way to tell it to build the needed packages before trying to run the dev server if they haven't already been built?

Is there a similar step for build and deploy too?  I was trying to test on Vercel and now that I've split some packages from the main app it isn't building them and thus errors.

**Comment by mmJordanO** [comment]
Update:  You might need a `prepare` script in your root level's package.json as well as some scoping on each package.

```
"build:package-a": "turbo run build --scope=package-a",
"build:package-b": "turbo run build --scope=package-b",
"prepare": "yarn build:package-a && yarn build:package-b"
```

This seems to make `yarn install` build the packages before trying to run the app.

**Reply by NullVoxPopuli** [reply]
This seems like the manual, non-turbo way to do things? This won't scale across 150 projects 🤔

**Reply by mmJordanO** [reply]
Good point.  I suppose once you get to that level of packages then detecting what needs building will be favorable.

Hopefully someone else knows a cleaner way to do this.

**Comment by baptisteArno** [comment]
Make sure your `app`'s `package.json` has `package-a: "*"` as a dependency?

**Reply by NullVoxPopuli** [reply]
I confirm I have * as the version(s)

**Comment by boennemann** [comment]
Have you tried

```
yarn turbo run start --scope="app" --include-dependencies
```

I found `--include-dependencies` to be necessary for scoped runs.

**Reply by NullVoxPopuli** [reply]
I have a bit of a sidequest to resolve, apparently: https://github.com/vercel/turborepo/issues/394

**Reply by NullVoxPopuli** [reply]
I tested this in an open source repo: https://github.com/NullVoxPopuli/limber/pull/377
(because I haven't figured out what's causing #394 yet)

and it doesn't quite work.
For example, if _any_ of my dependencies have a `start` script, that will still be executed (I don't want it to be)
![image](https://user-images.githubusercontent.com/199018/147258600-b2df3f95-3941-44a0-b270-83427eaaafad.png)

**Reply by NullVoxPopuli** [reply]
prune confirms that the correct packages are detected, so there's that:
```
❯ yarn turbo prune --scope="limber"
yarn run v1.22.17
$ /✂️/limber/node_modules/.bin/turbo prune --scope=limber
Generating pruned monorepo for limber in /✂️/limber/out
 - Added limber
 - Added @nullvoxpopuli/limber-monaco
 - Added @nullvoxpopuli/limber-codemirror
 - Added @nullvoxpopuli/horizon-theme
Done in 1.96s.
```

**Reply by NullVoxPopuli** [reply]
opened: https://github.com/vercel/turborepo/issues/417

**Reply by anthonyshew** [reply, marked as answer]
#417 was merged many moons ago so will close here!

TL;DR: Use the `--filter` flag!
